### PR TITLE
Loader 2 with failure fallback functionality

### LINKED ISSUE #529: ZiplineLoader Part 2
The ZiplineLoader has been wildly successful; let’s heap more responsibilities onto it.

### loadOrFallBack()

Attempt to fetch code from the network, fallback to the most-recently successful launched code, fallback again to the embedded code. This will require us to pin code that successfully launches (new persistent state to track!).

The DB should track multiple applications that share a cache. For example, we might track a bowling application and a golf application in the same Zipline project. Each gets its own pinned code and embedded manifest.

### loadContinuously()

Load only from the network, and produce a `Flow<Zipline>` every time the code changes. This is useful in development only. It should never fallback. If there’s nothing to load, the flow doesn’t emit.

### Constructing Ziplines

Currently `ZiplineLoader` takes a `Zipline` instance and loads code into it. We’ll move the responsibility of constructing the `Zipline` instance into `ZiplineLoader`. This should give us more flexibility and a better API.

### Loader + EventListener

We’ll add events to `EventListener` to track lifecycle events in ZiplineLoader:

```
  fun downloadStart(app: String, url: String)
  fun downloadEnd(app: String, url: String)
  fun downloadFailed(app: String, url: String, exception: IOException)

  fun applicationLoadStart(app: String, manifestUrl: String)
  fun applicationLoadEnd(app: String, manifestUrl: String)
  fun applicationLoadFailed(app: String, manifestUrl: String, exception: Exception)
```

These event functions are only ever called by the `zipline-loader` module, but they’re defined in the main `zipline` module.

### PR DESCRIPTION
**Pull request by adrw**
[CLOSES #529]

### CODE REVIEW

File: zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt [hunk 1]
```diff
@@ -37,10 +38,12 @@ import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.cancel
 import kotlinx.coroutines.isActive
+import kotlinx.serialization.ExperimentalSerializationApi
 import kotlinx.serialization.json.Json
 import kotlinx.serialization.modules.EmptySerializersModule
 import kotlinx.serialization.modules.SerializersModule
 
+@OptIn(ExperimentalSerializationApi::class)
```
Comment (swankjesse) [inline zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt:46]: Were we getting a warning about this before? Weird to introduce a warning suppression without introducing any API calls

File: zipline/src/commonMain/kotlin/app/cash/zipline/EventListener.kt [hunk 2]
```diff
@@ -67,6 +67,37 @@ abstract class EventListener {
   open fun serviceLeaked(name: String) {
   }
 
+  /** Invoked when an application load starts */
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/EventListener.kt:70]: Supernit on these: add a `.` to end the sentences!

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt [hunk 3]
```diff
@@ -48,18 +52,24 @@ class EmojiSearchZipline(
...
         cache = createZiplineCache(
+          eventListener = eventListener,
           driver = driver,
           fileSystem = FileSystem.SYSTEM,
           directory = cacheDir,
           nowMs = { Clock.systemDefaultZone().instant().toEpochMilli() },
         ),
-        delegate = HttpFetcher(httpClient = OkHttpZiplineHttpClient(client)),
+        delegate = HttpFetcher(
+          eventListener = eventListener, httpClient = OkHttpZiplineHttpClient(client)
```
Comment (swankjesse) [inline samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt]: supernit: wrap after the first , ?

File: zipline-cli/src/main/kotlin/app/cash/zipline/cli/Download.kt [hunk 4]
```diff
@@ -16,26 +16,35 @@
...
 import kotlinx.coroutines.runBlocking
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.modules.EmptySerializersModule
 import okhttp3.OkHttpClient
 import okio.FileSystem
 import okio.Path.Companion.toOkioPath
 import picocli.CommandLine.Command
 import picocli.CommandLine.Option
 
+@OptIn(ExperimentalSerializationApi::class)
```
Comment (swankjesse) [inline zipline-cli/src/main/kotlin/app/cash/zipline/cli/Download.kt]: What’s experimental here?

File: zipline-loader/README.md [hunk 5]
```diff
@@ -4,6 +4,35 @@ Zipline Loader
...
+
+## Pinning
+
+When a manifest successfully loads, it is pinned as stable.
+
+The manifest file and any constituent ZiplineFiles are pinned which prevents the files from being pruned.
+
+The pinned manifest is not unpinned until a later successful load of a different manifest.
+
+There is only ever 1 manifest pinned at a time. New manifest pins atomically remove the prior pinned manifest.
```
Comment (swankjesse) [inline zipline-loader/README.md]: add a prefix, "For each application, "

Comment (swankjesse) [reply zipline-loader/README.md]: Or add something about how pinning applies to applications?

File: zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt (hunk 1 above)
Comment (adrw) [reply zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt:46]: They were showing up as warnings in the logs and in IntelliJ, happy to keep ignoring

File: zipline-loader/src/jvmTest/kotlin/app/cash/zipline/loader/ZiplineCacheTest.kt [hunk 6]
```diff
@@ -171,6 +189,98 @@ class ZiplineCacheTest {
     }
   }
 
+  @Test
+  fun `new files are optimistically pinned`(): Unit = runBlocking {
```
Comment (swankjesse) [inline zipline-loader/src/jvmTest/kotlin/app/cash/zipline/loader/ZiplineCacheTest.kt:193]: awesome

File: zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt (hunk 1 above)
Comment (swankjesse) [reply zipline/src/engineMain/kotlin/app/cash/zipline/Zipline.kt:46]: I added a commit to fix.

File: zipline/src/commonMain/kotlin/app/cash/zipline/EventListener.kt (hunk 2 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/EventListener.kt:70]: I added a commit to fix.

File: samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt (hunk 3 above)
Comment (swankjesse) [reply samples/emoji-search/presenters/src/jvmMain/kotlin/app/cash/zipline/samples/emojisearch/EmojiSearchZipline.kt]: I added a commit to fix.

File: zipline-cli/src/main/kotlin/app/cash/zipline/cli/Download.kt (hunk 4 above)
Comment (swankjesse) [reply zipline-cli/src/main/kotlin/app/cash/zipline/cli/Download.kt]: I added a commit to fix.

File: zipline-loader/README.md (hunk 5 above)
Comment (swankjesse) [reply zipline-loader/README.md]: I added a commit to fix.

**Comment by swankjesse** [comment]
```
    e: /home/runner/work/zipline/zipline/zipline-gradle-plugin/src/test/projects/basic/lib/src/jvmMain/kotlin/app/cash/zipline/tests/launchGreetServiceJvm.kt: (42, 16): None of the following functions can be called with the arguments supplied: 
    public constructor ZiplineLoader(dispatcher: CoroutineDispatcher, eventListener: EventListener, serializersModule: SerializersModule, httpClient: ZiplineHttpClient, fetchers: List<Fetcher> = ...) defined in app.cash.zipline.loader.ZiplineLoader
    public fun ZiplineLoader(dispatcher: CoroutineDispatcher, serializersModule: SerializersModule, eventListener: EventListener, httpClient: OkHttpClient, fetchers: List<Fetcher> = ...): ZiplineLoader defined in app.cash.zipline.loader
    e: /home/runner/work/zipline/zipline/zipline-gradle-plugin/src/test/projects/basic/lib/src/jvmMain/kotlin/app/cash/zipline/tests/launchGreetServiceJvm.kt: (45, 23): None of the following functions can be called with the arguments supplied: 
    public constructor HttpFetcher(eventListener: EventListener, httpClient: ZiplineHttpClient) defined in app.cash.zipline.loader.fetcher.HttpFetcher
    public fun HttpFetcher(httpClient: OkHttpClient, eventListener: EventListener): HttpFetcher defined in app.cash.zipline.loader.fetcher
```

**Review by swankjesse — APPROVED** [review]

**Pull request merged**